src/realm/sync/noinst/client_reset.cpp

```diff
@@ -353,111 +461,115 @@ void client_reset::transfer_group(const Transaction& group_src, Transaction& gro
+                    col_key.is_dictionary() != col_key_dst.is_dictionary() ||
+                    col_key.is_set() != col_key_dst.is_set()) {
+                    throw ClientResetFailed(
+                        util::format("Incompatable column type change detected during client reset for '%1.%2'",
+                                     table_name, col_name));
+                }
+            }
         }
     }
 
     // Now the schemas are identical.
 
     // Remove objects in dst that are absent in src.
-    // We will also have to remove all objects created locally as they should have
-    // new keys because the client file id is changed.
-    auto new_file_id = group_dst.get_sync_file_id();
     for (auto table_key : group_src.get_table_keys()) {
+        if (!group_src.table_is_public(table_key))
+            continue;
         auto table_src = group_src.get_table(table_key);
         StringData table_name = table_src->get_name();
-        if (!table_name.begins_with("class"))
-            continue;
         logger.debug("Removing objects in '%1'", table_name);
         auto table_dst = group_dst.get_table(table_name);
-        if (auto pk_col = table_dst->get_primary_key_column()) {
-            std::vector<std::pair<Mixed, ObjKey>> objects_to_remove;
-            for (auto obj : *table_dst) {
-                auto pk = obj.get_any(pk_col);
-                if (!table_src->find_primary_key(pk)) {
-                    objects_to_remove.emplace_back(pk, obj.get_key());
-                }
-            }
-            for (auto& pair : objects_to_remove) {
-                logger.debug("  removing '%1'", pair.first);
-                table_dst->remove_object(pair.second);
+
+        auto pk_col = table_dst->get_primary_key_column();
+        REALM_ASSERT_DEBUG(pk_col); // sync realms always have a pk
+        std::vector<std::pair<Mixed, ObjKey>> objects_to_remove;
+        for (auto obj : *table_dst) {
+            auto pk = obj.get_any(pk_col);
+            if (!table_src->find_primary_key(pk)) {
+                objects_to_remove.emplace_back(pk, obj.get_key());
             }
         }
-        else {
-            std::vector<std::pair<GlobalKey, ObjKey>> objects_to_remove;
-            for (auto obj : *table_dst) {
-                auto oid = table_dst->get_object_id(obj.get_key());
-                auto key_src = table_src->get_objkey(oid);
-                if (oid.hi() == new_file_id || !key_src || !table_src->is_valid(key_src))
-                    objects_to_remove.emplace_back(oid, obj.get_key());
-            }
-            for (auto& pair : objects_to_remove) {
-                logger.debug("  removing '%1'", pair.first);
-                table_dst->remove_object(pair.second);
-            }
+        for (auto& pair : objects_to_remove) {
+            logger.debug("  removing '%1'", pair.first);
+            table_dst->remove_object(pair.second);
         }
     }
 
     // Add objects that are present in src but absent in dst.
```

Comment:
We could omit this step and just create the objects in the section below if they are not present in dest.

Reply:
I've updated to create objects as needed, instead of creating all objects beforehand in a separate loop. This improves performance by about 10% for the case where "remote will double the amount of local data". The other benchmarks are about the same.

<details>

Filters: client reset seamless loss

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
object-store-benchmarks is a Catch v2.13.3 host application.
Run with -? for options

-------------------------------------------------------------------------------
client reset seamless loss
  empty
-------------------------------------------------------------------------------
../test/object-store/benchmarks/client_reset.cpp:176
...............................................................................

benchmark name                       samples       iterations    estimated
                                     mean          low mean      high mean
                                     std dev       low std dev   high std dev
-------------------------------------------------------------------------------
no changes                                     100             2      4.252 ms
                                        21.9951 us    20.6819 us    24.3307 us
                                         8.6646 us    5.70202 us    13.2921 us


-------------------------------------------------------------------------------
client reset seamless loss
  populated with 10000 simple objects
  no change
-------------------------------------------------------------------------------
../test/object-store/benchmarks/client_reset.cpp:262
...............................................................................

benchmark name                       samples       iterations    estimated
                                     mean          low mean      high mean
                                     std dev       low std dev   high std dev
-------------------------------------------------------------------------------
reset with no changes                          100             1     2.96444 s
                                        24.3958 ms     24.045 ms     24.792 ms
                                        1.89776 ms    1.62387 ms    2.52848 ms


-------------------------------------------------------------------------------
client reset seamless loss
  populated with 10000 simple objects
  remote removes half the local data
-------------------------------------------------------------------------------
../test/object-store/benchmarks/client_reset.cpp:268
...............................................................................

benchmark name                       samples       iterations    estimated
                                     mean          low mean      high mean
                                     std dev       low std dev   high std dev
-------------------------------------------------------------------------------
reset will remove half the local
data                                           100             1       3.477 s
                                        35.6097 ms    35.2805 ms    35.9144 ms
                                         1.6163 ms    1.43266 ms    1.82333 ms


-------------------------------------------------------------------------------
client reset seamless loss
  populated with 10000 simple objects
  remote will double the amount of local data
-------------------------------------------------------------------------------
../test/object-store/benchmarks/client_reset.cpp:277
...............................................................................

benchmark name                       samples       iterations    estimated
                                     mean          low mean      high mean
                                     std dev       low std dev   high std dev
-------------------------------------------------------------------------------
reset will double the amount of
local data                                     100             1     4.98677 s
                                        47.7434 ms    47.2851 ms    48.1922 ms
                                        2.30836 ms    2.06076 ms    2.62881 ms


-------------------------------------------------------------------------------
client reset seamless loss
  with 5000 source objects linked to 5000 dest objects
  no change
-------------------------------------------------------------------------------
../test/object-store/benchmarks/client_reset.cpp:294
...............................................................................

benchmark name                       samples       iterations    estimated
                                     mean          low mean      high mean
                                     std dev       low std dev   high std dev
-------------------------------------------------------------------------------
reset with no changes                          100             1     3.07545 s
                                        30.9173 ms     30.581 ms    31.2689 ms
                                        1.75677 ms    1.53535 ms    2.13103 ms


-------------------------------------------------------------------------------
client reset seamless loss
  with 5000 source objects linked to 5000 dest objects
  remote removes half the local data
-------------------------------------------------------------------------------
../test/object-store/benchmarks/client_reset.cpp:300
...............................................................................

benchmark name                       samples       iterations    estimated
                                     mean          low mean      high mean
                                     std dev       low std dev   high std dev
-------------------------------------------------------------------------------
reset will remove half the local
data                                           100             1     4.58347 s
                                        46.3389 ms    45.9166 ms    46.7619 ms
                                        2.15682 ms    1.93457 ms    2.52667 ms


-------------------------------------------------------------------------------
client reset seamless loss
  with 5000 source objects linked to 5000 dest objects
  remote will double the amount of local data
-------------------------------------------------------------------------------
../test/object-store/benchmarks/client_reset.cpp:310
...............................................................................

benchmark name                       samples       iterations    estimated
                                     mean          low mean      high mean
                                     std dev       low std dev   high std dev
-------------------------------------------------------------------------------
reset will double the amount of
local data                                     100             1     6.66336 s
                                        64.4755 ms    63.9258 ms    65.0334 ms
                                        2.81509 ms    2.52643 ms    3.19359 ms


===============================================================================
All tests passed (38 assertions in 1 test case)